src/native/eventpipe/ds-dump-protocol.c

```diff
@@ -93,6 +93,71 @@ dump_protocol_helper_unknown_command (
 	return true;
 }
 
+static
+bool
+dump_protocol_generate_core_dump_response_flatten (
+	void *payload,
+	uint8_t **buffer,
+	uint16_t *size)
+{
+    DiagnosticsGenerateCoreDumpResponsePayload *response = (DiagnosticsGenerateCoreDumpResponsePayload*)payload;
+
+	EP_ASSERT (payload != NULL);
+	EP_ASSERT (buffer != NULL);
+	EP_ASSERT (*buffer != NULL);
+	EP_ASSERT (size != NULL);
+
+	bool success = true;
+
+	// int32_t error
+	memcpy (*buffer, &response->error, sizeof (response->error));
+	*buffer += sizeof (response->error);
+	*size -= sizeof (response->error);
+
+	// LPCWSTR error message
+	success &= ds_ipc_message_try_write_string_utf16_t (buffer, size, response->error_message);
+
+	// Assert we've used the whole buffer we were given
+	EP_ASSERT(*size == 0);
+
+    return success;
+}
+
+static
+void
+dump_protocol_generate_core_dump_response (
+	DiagnosticsIpcStream *stream,
+	ds_ipc_result_t error,
+    ep_char8_t * errorText)
+{
+    DiagnosticsGenerateCoreDumpResponsePayload payload;
+	DiagnosticsIpcMessage message;
+	ds_ipc_message_init (&message);
+
+    // Initialize payload
+    payload.error = error;
+    payload.error_message = ep_rt_utf8_to_utf16_string (errorText, -1);
+
+    // Get the size of the payload
+    size_t payload_size = sizeof(payload.error);
+	payload_size += sizeof(uint32_t);
+    payload_size += (payload.error_message != NULL) ? (ep_rt_utf16_string_len (payload.error_message) + 1) * sizeof(ep_char16_t) : 0;
+
+    // Initialize response message
+	bool result = ds_ipc_message_initialize_buffer (
+        &message,
+        ds_ipc_header_get_generic_error (),
+        &payload,
+        (uint16_t)payload_size,
+        dump_protocol_generate_core_dump_response_flatten);
+
+	if (result)
+		ds_ipc_message_send (&message, stream);
+
+	ds_ipc_message_fini (&message);
+	ep_rt_utf16_string_free ((ep_char16_t*)payload.error_message);
```

Comment:
If the struct should be owner of string, put the free of error_message in the fini function, if not, still call fini and then free the local owning the string, that will make the ownership of the string much clearer.